### ISSUE TITLE
proposal: "machine/sync" package

### ISSUE THREAD

**Issue by jadr2ddude** [tinygo-org/tinygo#712]
In order to use interrupts to implement wakeup from async operations, we need some high-level synchronization primitives to be available to the user. In the #606 experiment, a condition variable was added to a new package called `machine/sync`, as can be seen [here](https://github.com/jadr2ddude/tinygo/blob/d2b148340fd06f289943cb6acd205a47c5392222/src/machine/sync/cond.go).

Some remaining questions for this are:
1. What primitives are needed for common use cases?
2. What should this package be called (`machine/sync` was the first thing that came to mind, but it isn't necessarily that great)?

**Comment by deadprogram** [comment]
What about just using aka actually implementing the standard `sync`? Do we need a separate package?

**Comment by jadr2ddude** [comment]
This is different, so yes we need a separate package.

**Comment by aykevl** [comment]
I would propose `runtime/sync`, or maybe even `runtime/interrupt`. The reason for putting it under the interrupt package is because I'd like to keep special cases for the compiler limited to the runtime package as much as possible.

**Comment by jaddr2line** [comment]
In #1011, I went with `runtime/sync` as suggested

**Comment by deadprogram** [comment]
Now that #1142 has been merged, I think we can mark this issue as "next release". Is this correct @jaddr2line ?

**Comment by niaow** [comment]
No, I think this issue is more closely related to interrupt-safe condition variables which it seems are going to need to be held back to v0.15

**Comment by niaow** [comment]
#1172 dealt with this but I forgot about this issue.
